src/LocalDatastore.js

```diff
@@ -38,44 +56,69 @@ const LocalDatastore = {
     return controller.getAllContents();
   },
 
+  // Use for testing
+  _getRawStorage(): Promise {
+    const controller = CoreManager.getLocalDatastoreController();
+    return controller.getRawStorage();
+  },
+
   _clear(): Promise {
     const controller = CoreManager.getLocalDatastoreController();
     return controller.clear();
   },
 
   // Pin the object and children recursively
   // Saves the object and children key to Pin Name
-  async _handlePinWithName(name: string, object: ParseObject): Promise {
+  async _handlePinAllWithName(name: string, objects: Array<ParseObject>): Promise {
+    if (!objects || objects.length === 0) {
+      return;
+    }
     const pinName = this.getPinName(name);
-    const objects = this._getChildren(object);
-    objects[this.getKeyForObject(object)] = object._toFullJSON();
-    for (const objectKey in objects) {
-      await this.pinWithName(objectKey, objects[objectKey]);
+    const promises = [];
+    const objectKeys = [];
```

Comment:
def not asking you to do it, but this would be a good use for reduce....

Reply:
Just out of curiosity how would you do it with reduce :)

Reply:
is this to the point:

https://runkit.com/arthur/reduce-example